File: src/world/kernel.ts
import type {
  SimulationBridge,
  WorldAgentState,
  WorldCommand,
  WorldCommandResult,
  WorldEnvironmentState,
  WorldEventEnvelope,
  WorldStateBus,
  WorldStateSnapshot,
} from './types';

function nowIso(): string {
  return new Date().toISOString();
}

function createEnvironmentState(mode = 'sovereign'): WorldEnvironmentState {
  return {
    mode,
    regions: {},
    signals: {},
    updatedAt: nowIso(),
  };
}

function cloneAgent(agent: WorldAgentState): WorldAgentState {
  return {
    ...agent,
    position: agent.position ? { ...agent.position } : undefined,
    metrics: { ...agent.metrics },
    memory: [...agent.memory],
    tags: [...agent.tags],
  };
}

function cloneEvent(event: WorldEventEnvelope): WorldEventEnvelope {
  return {
    ...event,
    payload: { ...event.payload },
    causalityChain: [...event.causalityChain],
  };
}

function cloneAnomaly(anomaly: WorldStateSnapshot['anomalies'][number]): WorldStateSnapshot['anomalies'][number] {
  return {
    ...anomaly,
    causalityChain: [...anomaly.causalityChain],
  };
}

export class SovereignWorldKernel {
  private tick = 0;
  private status: WorldStateSnapshot['status'] = 'idle';
  private readonly worldId: string;
  private readonly metadata: Record<string, unknown>;
  private readonly agents = new Map<string, WorldAgentState>();
  private readonly anomalies: WorldStateSnapshot['anomalies'] = [];
  private readonly eventLog: WorldEventEnvelope[] = [];
  private environment: WorldEnvironmentState;

  constructor(
    private readonly stateBus: WorldStateBus,
    private readonly simulationBridge: SimulationBridge,
    options?: {
      worldId?: string;
      mode?: string;
      metadata?: Record<string, unknown>;
      initialSnapshot?: WorldStateSnapshot;
      initialEvents?: WorldEventEnvelope[];
    }
  ) {
    const initialSnapshot = options?.initialSnapshot;

    this.worldId = initialSnapshot?.worldId || options?.worldId || 'ionirix-sovereign-world';
    this.environment = initialSnapshot
      ? {
          ...initialSnapshot.environment,
          regions: { ...initialSnapshot.environment.regions },
          signals: { ...initialSnapshot.environment.signals },
        }
      : createEnvironmentState(options?.mode);
    this.metadata = {
      ...initialSnapshot?.metadata,
      ...options?.metadata,
      createdAt: (initialSnapshot?.metadata?.createdAt as string | undefined) || nowIso(),
      bridgeCapabilities: simulationBridge.getCapabilities(),
    };

    if (initialSnapshot) {
      this.tick = initialSnapshot.tick;
      this.status = initialSnapshot.status;

      for (const [agentId, agent] of Object.entries(initialSnapshot.agents)) {
        this.agents.set(agentId, cloneAgent(agent));
      }

      this.anomalies.splice(0, this.anomalies.length, ...initialSnapshot.anomalies.map((anomaly) => cloneAnomaly(anomaly)));
      this.eventLog.splice(
        0,
        this.eventLog.length,
        ...(options?.initialEvents || initialSnapshot.lastEvents).map((event) => cloneEvent(event))
      );
    }
  }

  getSnapshot(): WorldStateSnapshot {
    return this.buildSnapshot();
  }

  getEventLog(limit?: number): WorldEventEnvelope[] {
    const events = limit ? this.eventLog.slice(-limit) : this.eventLog;
    return events.map((event) => cloneEvent(event));
  }

  async pause(reason = 'manual'): Promise<WorldStateSnapshot> {
    this.status = 'paused';
    this.metadata.lastLifecycleAction = { action: 'pause', reason, at: nowIso() };
    this.eventLog.push(
      this.createEvent('world.lifecycle.paused', 'system', 'high', {
        reason,
      })
    );
    const snapshot = this.buildSnapshot();
    await this.stateBus.publishSnapshot(snapshot);
    return snapshot;
  }

  async resume(reason = 'manual'): Promise<WorldStateSnapshot> {
    this.status = 'running';
    this.metadata.lastLifecycleAction = { action: 'resume', reason, at: nowIso() };
    this.eventLog.push(
      this.createEvent('world.lifecycle.resumed', 'system', 'high', {
        reason,
      })
    );
    const snapshot = this.buildSnapshot();
    await this.stateBus.publishSnapshot(snapshot);
    return snapshot;
  }

  async persist(reason = 'manual'): Promise<WorldStateSnapshot> {
    const priorStatus = this.status;
    this.status = 'persisting';
    const persistedAt = nowIso();
    this.metadata.lastPersistedAt = persistedAt;
    this.metadata.lastLifecycleAction = { action: 'persist', reason, at: persistedAt };
    this.eventLog.push(
      this.createEvent('world.lifecycle.persisted', 'system', 'high', {
        reason,
        persistedAt,
      })
    );
    this.status = priorStatus === 'error' ? 'error' : priorStatus === 'paused' ? 'paused' : 'idle';
    const snapshot = this.buildSnapshot();
    await this.stateBus.publishSnapshot(snapshot);
    return snapshot;
  }

  async execute(command: WorldCommand): Promise<WorldCommandResult> {
    if (this.status === 'paused') {
      throw new Error('World kernel is paused.');
    }
    this.status = 'running';
    const emittedEvents: WorldEventEnvelope[] = [];
    let bridgeLifecycle: WorldStateSnapshot['status'] | undefined;

    if (command.type === 'spawn_agent') {
      const agent: WorldAgentState = {
        ...command.agent,
        updatedAt: command.agent.updatedAt || nowIso(),
      };
      this.agents.set(agent.id, cloneAgent(agent));
      emittedEvents.push(
        this.createEvent('world.agent.spawned', 'agent', 'normal', {
          agentId: agent.id,
          kind: agent.kind,
        })
      );
    }

    if (command.type === 'inject_event') {
      emittedEvents.push(
        this.createEvent(command.event.type, command.event.channel, command.event.priority, command.event.payload, {
          source: command.event.source,
          causalityChain: command.event.causalityChain,
        })
      );
    }

    if (command.type === 'modify_environment') {
      this.environment = {
        mode: command.patch.mode || this.environment.mode,
        regions: command.patch.regions || this.environment.regions,
        signals: command.patch.signals || this.environment.signals,
        updatedAt: nowIso(),
      };
      emittedEvents.push(
        this.createEvent('world.environment.modified', 'environment', 'high', {
          mode: this.environment.mode,
          metadata: command.patch.metadata || {},
        })
      );
    }

    if (command.type === 'run_scenario') {
      emittedEvents.push(
        this.createEvent('world.scenario.started', 'system', 'high', {
          scenarioId: command.scenarioId,
          directives: command.directives,
          targetTicks: command.targetTicks || 1,
        })
      );
    }

    const steps =
      command.type === 'advance_tick'
        ? Math.max(1, command.steps || 1)
        : command.type === 'run_scenario'
          ? Math.max(1, command.targetTicks || 1)
          : 0;
    for (let index = 0; index < steps; index += 1) {
      this.tick += 1;
      const bridgeResponse = await this.simulationBridge.advance({
        tick: this.tick,
        snapshot: this.buildSnapshot(),
        pendingEvents: emittedEvents,
      });

      if (bridgeResponse.agents) {
        for (const agent of Object.values(bridgeResponse.agents)) {
          this.agents.set(agent.id, cloneAgent({ ...agent, updatedAt: nowIso() }));
        }
      }

      if (bridgeResponse.environmentPatch) {
        this.environment = {
          ...this.environment,
          ...bridgeResponse.environmentPatch,
          regions: bridgeResponse.environmentPatch.regions || this.environment.regions,
          signals: bridgeResponse.environmentPatch.signals || this.environment.signals,
          updatedAt: nowIso(),
        };
      }

      if (bridgeResponse.anomalies?.length) {
        this.anomalies.splice(0, this.anomalies.length, ...bridgeResponse.anomalies);
      }

      if (bridgeResponse.metadata) {
        Object.assign(this.metadata, bridgeResponse.metadata);
      }

      if (bridgeResponse.lifecycle) {
        bridgeLifecycle = bridgeResponse.lifecycle;
      }

      if (bridgeResponse.events?.length) {
        emittedEvents.push(...bridgeResponse.events.map((event) => cloneEvent(event)));
      }
    }

    for (const event of emittedEvents) {
      this.eventLog.push(cloneEvent(event));
      await this.stateBus.publishEvent(event);
    }

  this.status = bridgeLifecycle && bridgeLifecycle !== 'running' ? bridgeLifecycle : 'idle';
    const snapshot = this.buildSnapshot();
    await this.stateBus.publishSnapshot(snapshot);

    return {
      accepted: true,
      commandType: command.type,
      tick: this.tick,
      snapshot,
      emittedEvents,
    };
  }

  private createEvent(
    type: string,
    channel: WorldEventEnvelope['channel'],
    priority: WorldEventEnvelope['priority'],
    payload: Record<string, unknown>,
    overrides?: {
      source?: string;
      causalityChain?: string[];
    }
  ): WorldEventEnvelope {
    return {
      id: crypto.randomUUID(),
      type,
      channel,
      priority,
      tick: this.tick,
      source: overrides?.source || 'sovereign-world-kernel',
      timestamp: nowIso(),
      payload,
      causalityChain: [...(overrides?.causalityChain || [])],
    };
  }

  private buildSnapshot(): WorldStateSnapshot {
    const version = `${this.worldId}:tick:${this.tick}`;
    const eventTail = this.eventLog.slice(-24).map((event) => cloneEvent(event));

    return {
      worldId: this.worldId,
      tick: this.tick,
      version,
      status: this.status,
      agents: Object.fromEntries([...this.agents.entries()].map(([key, agent]) => [key, cloneAgent(agent)])),
      environment: {
        ...this.environment,
        regions: { ...this.environment.regions },
        signals: { ...this.environment.signals },
      },
      anomalies: this.anomalies.map((anomaly) => cloneAnomaly(anomaly)),
      lastEvents: eventTail,
      frame: {
        frameId: crypto.randomUUID(),
        tick: this.tick,
        stateVersion: version,
        createdAt: nowIso(),
        eventIds: eventTail.map((event) => event.id),
        anomalyIds: this.anomalies.map((anomaly) => anomaly.id),
      },
      metadata: { ...this.metadata },
    };
  }
}

File: ION_ai/world/kernel.py
from __future__ import annotations

from dataclasses import asdict
from typing import Any, Dict, Iterable, Mapping

from ION_ai.simulation.engine import SimulationEngine

from .event_router import RoutedEvent, SovereignEventRouter
from .state import SovereignWorldState, WorldAnomaly, WorldSnapshot
from .tick_manager import TickManager

LIFECYCLE_INITIALIZING = "initializing"
LIFECYCLE_IDLE = "idle"
LIFECYCLE_RUNNING = "running"
LIFECYCLE_PAUSED = "paused"
LIFECYCLE_PERSISTING = "persisting"
LIFECYCLE_ERROR = "error"


class SovereignWorldKernel:
    def __init__(
        self,
        *,
        world_id: str = "ionirix-sovereign-world",
        simulation_engine: SimulationEngine | None = None,
        civilizations: Iterable[Any] | None = None,
        simulation_config: Mapping[str, Any] | None = None,
    ) -> None:
        self.tick_manager = TickManager()
        self.event_router = SovereignEventRouter()
        self.state = SovereignWorldState(world_id=world_id)
        self.simulation_engine = simulation_engine or SimulationEngine(
            civilizations=civilizations or [],
            config=simulation_config or {},
        )
        self.state.metadata.setdefault("lifecycle", [])
        self.state.metadata["kernel_contract"] = {
            "authoritative_runtime": "python",
            "event_router": "sovereign-event-router",
            "tick_manager": "authoritative",
        }

    def initialize(self, reason: str = "bootstrap") -> WorldSnapshot:
        self._transition_to(LIFECYCLE_INITIALIZING, reason=reason, emit_event=True)
        self.state.metadata["initialized_at"] = self.state.metadata.get("initialized_at") or self._utc_now_iso()
        self._transition_to(LIFECYCLE_IDLE, reason="kernel-ready", emit_event=True)
        return self.state.snapshot()

    def pause(self, reason: str = "manual") -> WorldSnapshot:
        self._transition_to(LIFECYCLE_PAUSED, reason=reason, emit_event=True)
        return self.state.snapshot()

    def resume(self, reason: str = "manual") -> WorldSnapshot:
        if self.state.status == LIFECYCLE_ERROR:
            raise RuntimeError("Cannot resume a kernel in error state.")
        self._transition_to(LIFECYCLE_RUNNING, reason=reason, emit_event=True)
        return self.state.snapshot()

    def fail(self, error_message: str) -> WorldSnapshot:
        self.state.metadata["last_error"] = error_message
        self._transition_to(LIFECYCLE_ERROR, reason=error_message, emit_event=True)
        return self.state.snapshot()

    def persist_snapshot(self, reason: str = "manual") -> WorldSnapshot:
        previous_status = self.state.status
        self._transition_to(LIFECYCLE_PERSISTING, reason=reason, emit_event=True)
        snapshot = self.state.snapshot()
        self.state.metadata["last_persisted_at"] = snapshot.frame.created_at
        self.state.metadata["last_persisted_version"] = snapshot.version
        restore_status = previous_status if previous_status != LIFECYCLE_PERSISTING else LIFECYCLE_IDLE
        self._transition_to(restore_status, reason="persist-complete", emit_event=True)
        return self.state.snapshot()

    def get_state_snapshot(self) -> WorldSnapshot:
        return self.state.snapshot()

    def spawn_agent(self, agent_id: str, payload: Mapping[str, Any]) -> WorldSnapshot:
        self._ensure_ready_for_mutation()
        self.state.upsert_agent(agent_id, payload)
        event = self.event_router.emit(
            event_type="world.agent.spawned",
            channel="agent",
            priority="normal",
            source="sovereign-world-kernel",
            tick=self.tick_manager.current_tick,
            payload={"agent_id": agent_id, **dict(payload)},
        )
        self._append_routed_event(event)
        return self.state.snapshot()

    def inject_event(self, payload: Mapping[str, Any]) -> RoutedEvent:
        self._ensure_ready_for_mutation()
        event = self.event_router.emit(
            event_type=str(payload.get("type", "world.external")),
            channel=str(payload.get("channel", "system")),
            priority=str(payload.get("priority", "normal")),
            source=str(payload.get("source", "external")),
            tick=self.tick_manager.current_tick,
            payload=dict(payload),
            causality_chain=list(payload.get("causality_chain", [])),
        )
        self._append_routed_event(event)
        return event

    def modify_environment(self, patch: Mapping[str, Any]) -> WorldSnapshot:
        self._ensure_ready_for_mutation()
        self.state.patch_environment(patch)
        event = self.event_router.emit(
            event_type="world.environment.modified",
            channel="environment",
            priority="high",
            source="sovereign-world-kernel",
            tick=self.tick_manager.current_tick,
            payload=dict(patch),
        )
        self._append_routed_event(event)
        return self.state.snapshot()

    def advance_tick(self, steps: int = 1, reason: str = "manual") -> WorldSnapshot:
        self._ensure_ready_for_tick()
        for _ in range(max(1, int(steps))):
            tick = self.tick_manager.advance(1, reason=reason)
            self.state.set_status(LIFECYCLE_RUNNING)
            self.state.advance_tick()
            simulation_output = self.simulation_engine.step()
            bridge_event = self.event_router.emit(
                event_type="world.tick.advanced",
                channel="kernel",
                priority="normal",
                source="sovereign-world-kernel",
                tick=tick,
                payload={
                    "reason": reason,
                    "pulse": simulation_output.get("pulse", {}),
                },
            )
            self._append_routed_event(bridge_event)
            anomalies = _extract_anomalies(simulation_output, tick=tick)
            self.state.replace_anomalies(anomalies)
            self.state.metadata["last_simulation_tick"] = simulation_output
            self.state.metadata["last_tick_reason"] = reason
        self.state.set_status(LIFECYCLE_IDLE)
        return self.state.snapshot()

    def run_scenario(self, scenario_id: str, directives: Iterable[str], target_ticks: int = 1) -> WorldSnapshot:
        self._ensure_ready_for_tick()
        scenario_event = self.event_router.emit(
            event_type="world.scenario.started",
            channel="system",
            priority="high",
            source="sovereign-world-kernel",
            tick=self.tick_manager.current_tick,
            payload={
                "scenario_id": scenario_id,
                "directives": list(directives),
                "target_ticks": target_ticks,
            },
        )
        self._append_routed_event(scenario_event)
        return self.advance_tick(steps=target_ticks, reason=f"scenario:{scenario_id}")

    def _ensure_ready_for_mutation(self) -> None:
        if self.state.status == LIFECYCLE_ERROR:
            raise RuntimeError("Kernel is in error state.")
        if self.state.status == LIFECYCLE_INITIALIZING:
            raise RuntimeError("Kernel is initializing.")
        if self.state.status == LIFECYCLE_PERSISTING:
            raise RuntimeError("Kernel is persisting state.")
        if self.state.metadata.get("initialized_at") is None:
            self.initialize()

    def _ensure_ready_for_tick(self) -> None:
        self._ensure_ready_for_mutation()
        if self.state.status == LIFECYCLE_PAUSED:
            raise RuntimeError("Kernel is paused.")

    def _append_routed_event(self, event: RoutedEvent) -> None:
        self.state.append_event(asdict(event))

    def _transition_to(self, status: str, *, reason: str, emit_event: bool) -> None:
        previous_status = self.state.status
        self.state.set_status(status)
        transition = {
            "from": previous_status,
            "to": status,
            "reason": reason,
            "tick": self.tick_manager.current_tick,
            "timestamp": self._utc_now_iso(),
        }
        self.state.metadata.setdefault("lifecycle", []).append(transition)
        self.state.metadata["last_status_change"] = transition

        if emit_event and previous_status != status:
            lifecycle_event = self.event_router.emit(
                event_type="world.lifecycle.changed",
                channel="system",
                priority="high" if status in {LIFECYCLE_ERROR, LIFECYCLE_PERSISTING} else "normal",
                source="sovereign-world-kernel",
                tick=self.tick_manager.current_tick,
                payload=transition,
                causality_chain=[f"lifecycle.{previous_status}", f"lifecycle.{status}"],
            )
            self._append_routed_event(lifecycle_event)

    def _utc_now_iso(self) -> str:
        return self.state.environment.get("updated_at", "") or self.state.snapshot().frame.created_at


def _extract_anomalies(simulation_output: Mapping[str, Any], *, tick: int) -> list[WorldAnomaly]:
    anomalies: list[WorldAnomaly] = []
    pulse = simulation_output.get("pulse", {})
    if isinstance(pulse, Mapping):
        tension = float(pulse.get("tension", 0.0))
        vitality = float(pulse.get("vitality", 1.0))
        if tension >= 0.85:
            anomalies.append(
                WorldAnomaly(
                    anomaly_id=f"anomaly-tension-{tick}",
                    anomaly_type="tension_spike",
                    severity="high",
                    summary="Global tension exceeded sovereign threshold.",
                    tick=tick,
                    causality_chain=["pulse.tension"],
                )
            )
        if vitality <= 0.2:
            anomalies.append(
                WorldAnomaly(
                    anomaly_id=f"anomaly-vitality-{tick}",
                    anomaly_type="vitality_collapse",
                    severity="critical",
                    summary="World vitality collapsed below resilience floor.",
                    tick=tick,
                    causality_chain=["pulse.vitality"],
                )
            )
        cooperation = float(pulse.get("cooperation", 1.0))
        if cooperation <= 0.15:
            anomalies.append(
                WorldAnomaly(
                    anomaly_id=f"anomaly-cooperation-{tick}",
                    anomaly_type="cooperation_breakdown",
                    severity="high",
                    summary="Cooperation collapsed below sovereign continuity threshold.",
                    tick=tick,
                    causality_chain=["pulse.cooperation"],
                )
            )

    existential = simulation_output.get("existential", {})
    crises = existential.get("crises", []) if isinstance(existential, Mapping) else []
    for index, crisis in enumerate(crises):
        if not isinstance(crisis, Mapping):
            continue
        severity = float(crisis.get("severity", 0.0))
        if severity < 0.75:
            continue
        crisis_type = str(crisis.get("risk_type", "systemic"))
        civilization = str(crisis.get("civilization", "unknown"))
        anomalies.append(
            WorldAnomaly(
                anomaly_id=f"anomaly-crisis-{tick}-{index}",
                anomaly_type=f"existential_{crisis_type}",
                severity="critical" if severity >= 0.9 else "high",
                summary=f"Existential crisis pressure exceeded threshold for {civilization}.",
                tick=tick,
                causality_chain=[f"existential.crises[{index}]", f"civilization.{civilization}"],
            )
        )

    strategy = simulation_output.get("grand_strategy", {})
    strategy_outcomes = strategy.get("outcomes", {}) if isinstance(strategy, Mapping) else {}
    if isinstance(strategy_outcomes, Mapping) and strategy_outcomes:
        composites = []
        for metrics in strategy_outcomes.values():
            if isinstance(metrics, Mapping):
                composites.append(float(metrics.get("composite", 0.0)))
        if composites and sum(composites) / len(composites) <= 0.2:
            anomalies.append(
                WorldAnomaly(
                    anomaly_id=f"anomaly-strategy-{tick}",
                    anomaly_type="strategy_stall",
                    severity="medium",
                    summary="Strategic posture degraded below coherent planning threshold.",
                    tick=tick,
                    causality_chain=["grand_strategy.outcomes"],
                )
            )
    return anomalies